### ISSUE TITLE
E1102 should not fire on fields in mock objects that have been given a return_value.

### ISSUE THREAD

**Issue by mkillianey** [pylint-dev/pylint#9216]
### Bug description

```python
When using unittest.mock.Mock, local changes to Mock objects seem to poison pylint's analysis for the general population of Mocks, causing incorrectly reported E1102 `not-callable` errors.

Also observed, dunno if related -- calls to `return_value` are incorrectly permitted on objects that are not Mock objects and do not have a `return_value` method.

Test input:  `example.py`


"""This is example.py."""

from unittest.mock import Mock
from dataclasses import dataclass
from typing import Callable


# If the following line is commented out, the problem goes away
Mock().example = 123


@dataclass
class _SomeClass:
    example: Callable[[int], int]


def _some_method(x: int) -> int:
    return x


def _unhinted_mock() -> None:
    unhinted_mock = Mock()
    unhinted_mock.example.return_value = 123
    unhinted_mock.example(123)


def _hinted_mock() -> None:
    hinted_as_mock: Mock = Mock()
    hinted_as_mock.example.return_value = 123
    hinted_as_mock.example(123)


def _mock_hinted_as_class() -> None:
    hinted_as_class: _SomeClass = Mock()
    hinted_as_class.example.return_value = 123
    hinted_as_class.example(123)


def _actual_class() -> None:
    thing: _SomeClass = _SomeClass(_some_method)
    thing.example.return_value = 123
    thing.example(123)
```

### Command used

```shell
pylint example.py
```

### Pylint output

```python
************* Module example
example.py:24:4: E1102: unhinted_mock.example is not callable (not-callable)
example.py:30:4: E1102: hinted_as_mock.example is not callable (not-callable)
example.py:36:4: E1102: hinted_as_class.example is not callable (not-callable)

------------------------------------------------------------------
Your code has been rated at 3.75/10 (previous run: 3.75/10, +0.00)
```

### Expected behavior

Either no errors, or perhaps one or both of the objects hinted as non-Mock classes should complain about the use of methods from Mock -- maybe `no-member`, like this?
```
example.py:35:4: E1101: Function 'hinted_as_class.example' has no 'return_value' member (no-member)
example.py:41:4: E1101: Function 'thing.example' has no 'return_value' member (no-member)
```

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.5 (main, Aug 25 2023, 13:19:53) [GCC 9.4.0]
```

### OS / Environment

Ubuntu 20.04

```
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.6 LTS"
$ uname -a
Linux <redacted> 5.15.0-83-generic #92~20.04.1-Ubuntu SMP Mon Aug 21 14:00:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

### Additional dependencies

```python
astroid==3.0.1
dill==0.3.7
isort==5.12.0
mccabe==0.7.0
platformdirs==3.11.0
pylint==3.0.2
```
